### ISSUE TITLE
[TextInput] Get keydown event to handle "Enter"

### ISSUE THREAD

**Issue by JonasJonny** [facebook/react-native#1882]
Hello.
I tried to catch keydown/keypress events for several hours with no luck.
I can get whole .text string (onChange={}) but what if I need to do an action when "Enter/Backspace" are pressed?
Is there any solution how to know which keyboard character was just pressed in TextInput keyboard?

Thank you.

**Comment by chirag04** [comment]
Not sure if ``controlled={true}`` can help you here.

**Comment by JonasJonny** [comment]
No luck with ``controlled={true}`` but thank you @chirag04.
Attributes onKeyDown/Up/Pressed didn't work but I know they are not supported. 
Event for onChange have just target, text, timestamp :(.

**Comment by brentvatne** [comment]
I would like to see `onKey*` implemented, @sahrens do you have any suggestions for where to get started on implementing this? Maybe @JonasJonny or another interested contributor would be able to take that and turn it into a PR

**Comment by JonasJonny** [comment]
Unfortunately I don't have enough skill to help.

**Comment by grabbou** [comment]
@brentvatne I think the way you emit events for your `video` component should work quite nicely in this case.

**Comment by dsibiski** [comment]
@JonasJonny I have a working implementation of `onKeyPress` that successfully detects `return` and `delete`.

Right now it is returning the character of the key that was pressed and for `return` and `delete`, it's returning ASCII codes `CR` and `DEL` respectively.

Would this take care of your use case? @brentvatne any other ideas as to how the `return` and `delete` codes should be returned to JS? Would it be better to use ASCII code numbers?

I'd like to figure these things out before submitting a PR.

**Comment by dsibiski** [comment]
Ah, it looks like there really isn't a `delete` key in iOS, it's more like a `backspace`, so it should send back `BS` or `08`

**Comment by JonasJonny** [comment]
Thank you @dsibiski for following. I really appreciate that.

In my opinion you should return "keyCode" http://www.cambiaresearch.com/articles/15/javascript-char-codes-key-codes

**Comment by brentvatne** [comment]
Agreed with @JonasJonny - I think we should emulate the modern browser version of this: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key - so it would return {key: "Backspace"} and [some other properties too](https://developer.mozilla.org/en-US/docs/Web/Events/keypress) (probably for now we can just do key and implement others on a per-use-case basis)

**Comment by dsibiski** [comment]
Thanks for the feedback guys! I pretty much have this ready for a PR, I'll make these changes and you guys can review it and see what you think...

**Comment by browniefed** [comment]
@brentvatne not sure how you want to handle issues w/ pending PRs.

**Comment by grabbou** [comment]
`PR submitted` label would've been handy when scrolling through the issues to pick one to work on.

**Comment by brentvatne** [comment]
@browniefed - yeah it's a little awkward, maybe PR submitted tag would be good but that might be hard to keep up to date unless it's added automatically

**Comment by ide** [comment]
GitHub should build this if there isn't already a way to write this query.

**Comment by browniefed** [comment]
@ide +1 ... oh wait they haven't built a voting system yet ... not holding my breath.

**Comment by GantMan** [comment]
Anything I can do to help this task along?   Pretty essential to native apps I believe.

**Comment by dsibiski** [comment]
@brentvatne We can close this now as my PR for `onKeyPress` has been merged to master.

**Comment by brentvatne** [comment]
Awesome :smile:

**Comment by yfsx** [comment]
what about android? is onKeyPress worked on android version? because it's not possible to make 2 versions of code only for "enter clicked".

**Comment by brentvatne** [comment]
cc @dsibiski ^

**Comment by Abhay-Joshi-Git** [comment]
Does Android version has support for onKeyPress ? Documentation suggests otherwise and I just tried and it's not working with version 0.26.2, but I just wanted to get this confirmed. 
Also, it would be great if anyone can estimate about when we would be able to see that if it's not already there.

**Comment by dsibiski** [comment]
@yfsx @Abhay-Joshi-Git: No, it hasn't yet been implemented for Android. If either of you want to give it a shot that would be awesome 👍

**Comment by semiautomatix** [comment]
Any further work on this? I am relying on a barcode scanner - that produces a return key value upon completion. If this isn't available for Android then it's a dealbreaker.

**Comment by santhosh77h** [comment]
Any update for android.`OnKeyPress`

**Comment by uncle-T0ny** [comment]
What the problem with https://github.com/facebook/react-native/issues/2082 ?

**Comment by joshyhargreaves** [comment]
Heads up: I'm taking a look at doing this on Android.

**Comment by joshyhargreaves** [comment]
Does anyone have any thoughts on the best way to approach this in terms of detecting the keypresses on ReactEditText? It seems that there might be a few approaches. 

**TextWatcher**
The least desirable approach would be use the TextWatcher interface to determine the key press based on a the difference between the oldText and the newText, [here](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java#L600) for example. However this really doesn't seem feasible, due to numerous reasons; difficulty detecting backspaces on empty text views for one?

**KeyListener**
In ReactEditText we are using an internal KeyListener, which perhaps could be a place where we could capture the keypress [here](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java#L656)? However the relevant onKeyDown/onKeyUpMethods methods only seem to fire when the keyboard type is not of type TYPE_CLASS_TEXT; i.e. only when the keyboard is numeric, for example. Any ideas as to why this might be, I'm guessing there's no need internally to call the KeyListener if the set input type is of type text? 

**InputConnectionWrapper**
Another approach seems to be extending the InputConnectionWrapper class and capturing the keypresses that way, for example as per [here](http://stackoverflow.com/questions/4886858/android-edittext-deletebackspace-key-event/16551670#16551670). I tried this out [here](https://github.com/facebook/react-native/compare/master...joshyhargreaves:feature/1882_implement_keycode_callback), but experienced the same behaviour as the KeyListener; the events were only firing if the keyboard shown was default/of class TEXT. 

@janicduplessis I'm not sure if you can offer any insight :)?

**Comment by GuillaumeMunsch** [comment]
Hi !

Any news on this ?

**Comment by vance-liu** [comment]
Hi ! e~~~
Any news on this ~~~?

**Comment by andela-landia** [comment]
Hi, you can use `onSubmitEditing` prop on the TextInput to submit on `Enter`. This works for iOS and Android.

**Comment by hkung77** [comment]
onSubmitEditing works great for Enter but how about listening for other input events like backspace?

**Comment by joshyhargreaves** [comment]
Hey everyone, 
I created a PR for `onKeyPress` on Android: thoughts and comments much appreciated :). 
https://github.com/facebook/react-native/pull/14720
![onkeypressandroid](https://user-images.githubusercontent.com/1807207/27512940-e2ae09ce-594a-11e7-9a82-6efa3e9c2f61.gif)
The PR itself has a fairly extensive comment on implementation details.

**Comment by semiautomatix** [comment]
@joshyhargreaves very excited for this PR - however I've changed my entire project to work around this shortcoming already.

**Comment by GuillaumeMunsch** [comment]
Looking forward to see this in RN next version as well ! :)

**Comment by joshyhargreaves** [comment]
So a bit of an update on the PR.

the tl:dr is that there are some limitations to the native APIs that stop us being able to have a completely 'flawless' `onKeyPress` API, but I reason that we don't need it to be a completely 'flawless ''onKeyPress' implementation, as implementing such a thing is something we cannot do natively. What we need to do is expose native functionality to JS that users currently don't have on Android, like detecting backspaces where no text changes, and distinguishing between text changes from keyboards and cut/paste. These are things we can do, and I'll update the PR to support these cases and bring the API functionality as close to the iOS API as possible. 

https://github.com/facebook/react-native/pull/14720#issuecomment-311192996

**Comment by bunnyc1986** [comment]
@joshyhargreaves, thanks for creating this PR. That's exact what I am looking for. 
May I ask what is the status? Is it going to be merged soon? I look forward to seeing it on next RN version.

**Comment by gianpaj** [comment]
Same question to @joshyhargreaves on the status on https://github.com/facebook/react-native/pull/14720

FYI this is feature request on Canny
https://react-native.canny.io/feature-requests/p/textinput-onkeypress-for-android

**Comment by joshyhargreaves** [comment]
@gianpaj this landed a while ago! https://twitter.com/joshyhargreaves/status/949022712814555142
